pyccl/pkresponse.py

```diff
@@ -0,0 +1,795 @@
+
+    # Growth factor
+    Dp = cosmo_hp.growth_factor_unnorm(a_arr)
+    Dm = cosmo_hm.growth_factor_unnorm(a_arr)
+
+    na = len(a_arr)
+    nk = len(k_use)
+    dpk12 = np.zeros([na, nk])
+    logMfor_hmf = np.linspace(8, 17, 200)
+    logMh = np.linspace(log10Mh_min, log10Mh_max, 2**5 + 1)  # M_sol/h
+    logM = np.log10(10**logMh / h)
+    Mh = 10**logMh
+    M = 10**logM
+    nM = len(M)
+    dlogM = logM[1] - logM[0]
+    b1_th_tink = np.zeros(nM)
+    Pth = np.zeros((nM, nk))
+    Pnth_hp = np.zeros((nM, nk))
+    Pnth_hm = np.zeros((nM, nk))
+    Pbin = np.zeros((nM, nk))
+
+    nths = np.zeros(nM)
+
+    mass_def = halos.MassDef200m
+    hmf = halos.MassFuncDarkEmulator(mass_def=mass_def)
+    hbf = halos.HaloBiasTinker10(mass_def=mass_def)
+
+    # dark emulator is valid for 0 =< z <= 1.48
+    if np.any(1.0 / a_arr - 1) > 1.5:
+        print("dark emulator is valid for z={:.2f}<1.48")
+
+    for ia, aa in enumerate(a_arr):
+        z = 1.0 / aa - 1
+
+        # mass function
+        dndlog10m_emu = ius(
+            logMfor_hmf, hmf(cosmo, 10**logMfor_hmf, aa)
+        )  # Mpc^-3
+
+        darkemu_set_cosmology(emu, cosmo)
+        for m in range(nM):
+            Pth[m] = emu.get_phm_massthreshold(k_emu, Mh[m], z) * (1 / h) ** 3
+            Pbin[m] = emu.get_phm_mass(k_emu, Mh[m], z) * (1 / h) ** 3
+            nths[m] = emu.mass_to_dens(Mh[m], z) * h**3
+
+            logM1 = np.linspace(logM[m], logM[-1], 2**5 + 1)
+            dlogM1 = logM[1] - logM[0]
+
+            b1_th_tink[m] = integrate.romb(
+                dndlog10m_emu(logM1) * hbf(cosmo, (10**logM1), aa), dx=dlogM1
+            ) / integrate.romb(dndlog10m_emu(logM1), dx=dlogM1)
+
+        darkemu_set_cosmology(emu, cosmo_hp)
+        for m in range(nM):
+            Pnth_hp[m] = (
+                emu.get_phm(
+                    k_emu * (h / hp), np.log10(nths[m] * (1 / hp) ** 3), z
+                )
+                * (1 / hp) ** 3
+            )
+
+        darkemu_set_cosmology(emu, cosmo_hm)
+        for m in range(nM):
+            Pnth_hm[m] = (
+                emu.get_phm(
+                    k_emu * (h / hm), np.log10(nths[m] * (1 / hm) ** 3), z
+                )
+                * (1 / hm) ** 3
+            )
+
+        Nc = prof_hod._Nc(M, aa)
+        Ns = prof_hod._Ns(M, aa)
+        fc = prof_hod._fc(aa)
+        Ng = Nc * (fc + Ns)
+        logMps = logM + dlogM
+        logMms = logM - dlogM
+
+        prof_Mp = prof_hod.fourier(cosmo, k_use, (10**logMps), aa)
+        prof_Mm = prof_hod.fourier(cosmo, k_use, (10**logMms), aa)
+
+        dprof_dlogM = (prof_Mp - prof_Mm) / (2 * dlogM)
+
+        nth_mat = np.tile(nths, (len(k_use), 1)).transpose()
+        ng = integrate.romb(dndlog10m_emu(logM) * Ng, dx=dlogM, axis=0)
+        bgE = (
+            integrate.romb(
+                dndlog10m_emu(logM) * Ng * (hbf(cosmo, M, aa)),
+                dx=dlogM,
+                axis=0,
+            )
+            / ng
+        )
+
+        bgE2 = (
```

Comment:
@damonge , can this be done with the current methods of the halo model?

Reply:
Same with the other integrals

Reply:
I think so. @RyoTerasawa can you check:
https://github.com/LSSTDESC/CCL/blob/5273ca6eb76d7c92fd11cde42bf02f761e8c29cd/pyccl/halos/halo_model.py#L108
and 
https://github.com/LSSTDESC/CCL/blob/5273ca6eb76d7c92fd11cde42bf02f761e8c29cd/pyccl/halos/halo_model.py#L113
and
https://github.com/LSSTDESC/CCL/blob/5273ca6eb76d7c92fd11cde42bf02f761e8c29cd/pyccl/halos/halo_model.py#L118
?

Reply:
I changed the `pkresponse.py`, and now all the integrals of the mass function are using `_integrate_over_mf` or `_integrate_over_mbf`.